spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -178,8 +178,12 @@ private static void apply(UpdateSchema pendingUpdate, TableChange.UpdateColumnPo
 
   private static void apply(UpdateSchema pendingUpdate, TableChange.AddColumn add) {
     Type type = SparkSchemaUtil.convert(add.dataType());
-    pendingUpdate.addColumn(parentName(add.fieldNames()), leafName(add.fieldNames()), type, add.comment());
-
+    if (add.isNullable()) {
+      pendingUpdate.addColumn(parentName(add.fieldNames()), leafName(add.fieldNames()), type, add.comment());
+    } else {
+      pendingUpdate.allowIncompatibleChanges()
```

Comment:
I'm not sure that this should call `allowIncompatibleChanges()` because adding a required column when there are no existing values will break reading the new column in any table with data in it. The only time it is safe to add a required column is if there is no data in the table.

What about throwing an exception here instead? I agree that the column should not be optional if NOT NULL was specified.

Another alternative is to check whether the table has data and allow the incompatible change if it doesn't have any rows.

Reply:
Thanks for the review, @rdblue.  Sure, that makes sense. 

I will update it to throw an error and will use the same error message as SchemaUpdate.addRequiredColumn would have returned without the `allowIncompatibleChanges()`. 
 ie ` Incompatible change: cannot add required column: %s`